File: day_19_interfaces/Day19.java

// declaring our interface
interface AdvancedArithmetic {
// declaring our abstract method
    int divisorSum(int n);
}

class Calculator implements AdvancedArithmetic {

// empty constructor
    Calculator() {}

// overriding the abstract method
    @Override
    public int divisorSum(int n) {
        int total = 1 + n;
        for (int i = 2; i < (n / 2) + 1; i++) {
            if (n % i == 0) {
                total = total + i;
            }
        }
        return total;
    }

    public static void main(String[] args) {
        Calculator c = new Calculator();
        System.out.println(c.divisorSum(6));
        System.out.println(c.divisorSum(10));
        System.out.println(c.divisorSum(55));
    }
}

File: day_19_interfaces/day_19.py
# Day 19

class AdvancedArithmetic(object):
    def divisorSum(n):
        raise NotImplementedError


class Calculator(AdvancedArithmetic):
    def divisorSum(self, n):
        divisors = [1, n]
        for i in range(2, (n//2) + 1):
            if n % i == 0:
                divisors.append(i)
        return sum(set(divisors))


c = Calculator()

c.divisorSum(55)


# alternative: using abcmeta will return an error when trying to instantiate a subclass not overriding an interface (abstract) method.
# from abc import ABCMeta, abstractmethod

# class AdvancedArithmetic(metaclass = ABCMeta):
#     @abstractmethod
#     def divisorSum(n):
#         raise NotImplementedError

# class Calculator(AdvancedArithmetic):
#     def divisorSum(self, n):
#         divisors = [1, n]
#         for i in range(2, (n//2) + 1):
#             if n % i == 0:
#                 divisors.append(i)
#         return sum(set(divisors))